File: jsrc/machine/SxxMachine/ErrorTerm.java
package com.googlecode.prolog_cafe.lang;

import java.util.IdentityHashMap;

/**
 * A term to represent errors, extends {@link StructureTerm}, 
 * additionally holds reference to {@link Throwable}, which this term represents.
 * @author a.semenov@unipro.ru
 *
 */
public class ErrorTerm extends StructureTerm {

	private final Throwable error;
	
	public ErrorTerm(Throwable error, String name, Term... args){
		super(name, args);
		this.error = error;
	}
	
	public ErrorTerm(Throwable error, SymbolTerm _functor, Term... _args){
		super(_functor, _args);
		this.error = error;
	}

	public Throwable getThrowable() {
		return error;
	}	
	
	@Override
	protected Term copy(IdentityHashMap<VariableTerm,VariableTerm> copyHash) {
		Term[] a = new Term[arity()];
		for (int i = 0; i < a.length; i++){
			a[i] = arg(i).copy(copyHash);
		}
		return new ErrorTerm(error, functor(), a);
	}
}

File: src/machine/SxxMachine/ErrorTerm.cpp
using namespace std;

#include "ErrorTerm.h"
#include "Term.h"
#include "TermData.h"

namespace SxxMachine
{

	bool ErrorTerm::isError()
	{
		return true;
	}

	ErrorTerm::ErrorTerm(runtime_error error, const wstring& name, vector<Term> &args) : StructureTerm(name, args), error(error)
	{
	}

	ErrorTerm::ErrorTerm(runtime_error error, Term* _functor, vector<Term> &_args) : StructureTerm(_functor, _args), error(error)
	{
	}

	runtime_error ErrorTerm::getThrowable()
	{
		return this->error;
	}

	Term* ErrorTerm::copyImpl(IdentityHashMap<any, Term*>* copyHash, const int& deeply)
	{
		std::vector<Term*> a(arity());
		for(int i = 0; i < a.size(); i++)
		{
			a[i] = arg0(i)->copy(copyHash, deeply);
		}
		return TermData::createErrorTerm(this->error, functor(), a);
	}
}
